engine/consensus/approvals/assignment_collector.go

```diff
@@ -0,0 +1,346 @@
+	}
+	return collectors
+}
+
+func (rsr *AssignmentCollector) verifyAttestationSignature(approval *flow.ResultApprovalBody, nodeIdentity *flow.Identity) error {
+	id := approval.Attestation.ID()
+	valid, err := rsr.verifier.Verify(id[:], approval.AttestationSignature, nodeIdentity.StakingPubKey)
+	if err != nil {
+		return fmt.Errorf("failed to verify attestation signature: %w", err)
+	}
+
+	if !valid {
+		return engine.NewInvalidInputErrorf("invalid attestation signature for (%x)", nodeIdentity.NodeID)
+	}
+
+	return nil
+}
+
+func (rsr *AssignmentCollector) verifySignature(approval *flow.ResultApproval, nodeIdentity *flow.Identity) error {
+	id := approval.Body.ID()
+	valid, err := rsr.verifier.Verify(id[:], approval.VerifierSignature, nodeIdentity.StakingPubKey)
+	if err != nil {
+		return fmt.Errorf("failed to verify signature: %w", err)
+	}
+
+	if !valid {
+		return engine.NewInvalidInputErrorf("invalid signature for (%x)", nodeIdentity.NodeID)
+	}
+
+	return nil
+}
+
+// validateApproval performs result level checks of flow.ResultApproval
+// checks:
+// - verification node identity
+// - attestation signature
+// - signature of verification node
+// - chunk index sanity check
+// - block ID sanity check
+// Returns:
+// - engine.InvalidInputError - result approval is invalid
+// - exception in case of any other error, usually this is not expected
+// - nil on successful check
+func (rsr *AssignmentCollector) validateApproval(approval *flow.ResultApproval) error {
+	// check that approval is for the expected result to reject incompatible inputs
+	if approval.Body.ExecutionResultID != rsr.ResultID {
+		return fmt.Errorf("this AssignmentCollector processes only approvals for result (%x) but got an approval for (%x)", rsr.ResultID, approval.Body.ExecutionResultID)
+	}
+
+	// approval has to refer same block as execution result
+	if approval.Body.BlockID != rsr.BlockID() {
+		return engine.NewInvalidInputErrorf("result approval for invalid block, expected (%x) vs (%x)",
+			rsr.BlockID(), approval.Body.BlockID)
+	}
+
+	chunkIndex := approval.Body.ChunkIndex
+	if chunkIndex >= uint64(rsr.result.Chunks.Len()) {
+		return engine.NewInvalidInputErrorf("chunk index out of range: %v", chunkIndex)
+	}
+
+	identity, found := rsr.authorizedApprovers[approval.Body.ApproverID]
+	if !found {
+		return engine.NewInvalidInputErrorf("approval not from authorized verifier")
+	}
+
+	err := rsr.verifyAttestationSignature(&approval.Body, identity)
+	if err != nil {
+		return fmt.Errorf("validating attestation signature failed: %w", err)
+	}
+
+	err = rsr.verifySignature(approval, identity)
+	if err != nil {
+		return fmt.Errorf("validating approval signature failed: %w", err)
+	}
+
+	return nil
+}
+
+func (rsr *AssignmentCollector) ProcessApproval(approval *flow.ResultApproval) error {
+	err := rsr.validateApproval(approval)
+	if err != nil {
+		return fmt.Errorf("could not validate approval: %w", err)
+	}
+
+	rsr.verifiedApprovalsCache.Put(approval)
```

Comment:
Would be great to first check with the cache, if the approval is already present. Only if it is _not_ present, we should validate it and subsequently add it. Otherwise, we can its a duplicate.